util/iostat/iostat.go

```diff
@@ -0,0 +1,132 @@
+package iostat
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"os/exec"
+	"strconv"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/metrics"
+)
+
+type MetricsSpawner struct {
+	statReceiver chan DeviceStats
+	metrics      map[string]map[string]metrics.GaugeFloat64
+}
+
+func NewMetricsSpawner() *MetricsSpawner {
+	return &MetricsSpawner{
+		metrics:      make(map[string]map[string]metrics.GaugeFloat64),
+		statReceiver: make(chan DeviceStats),
+	}
+}
+
+func (m *MetricsSpawner) RegisterMetrics(ctx context.Context, spwanInterval int) error {
+	go Run(ctx, spwanInterval, m.statReceiver)
+	// Register metrics for a maximum of 5 devices (fail safe incase iostat command returns incorrect names indefinitely)
+	for i := 0; i < 5; i++ {
+		stat, ok := <-m.statReceiver
+		if !ok {
+			return errors.New("failed to register iostat metrics")
+		}
+		if _, ok := m.metrics[stat.DeviceName]; ok {
+			return nil
+		}
+		baseMetricName := fmt.Sprintf("isotat/%s/", stat.DeviceName)
+		m.metrics[stat.DeviceName] = make(map[string]metrics.GaugeFloat64)
+		m.metrics[stat.DeviceName]["readspersecond"] = metrics.NewRegisteredGaugeFloat64(baseMetricName+"readspersecond", nil)
+		m.metrics[stat.DeviceName]["writespersecond"] = metrics.NewRegisteredGaugeFloat64(baseMetricName+"writespersecond", nil)
+		m.metrics[stat.DeviceName]["await"] = metrics.NewRegisteredGaugeFloat64(baseMetricName+"await", nil)
+	}
+	return nil
+}
+
+func (m *MetricsSpawner) PopulateMetrics() {
+	for {
+		stat, ok := <-m.statReceiver
+		if !ok {
+			log.Info("Iostat statReceiver channel was closed due to error or command being completed")
+			return
+		}
+		if _, ok := m.metrics[stat.DeviceName]; !ok {
+			log.Warn("Unrecognized device name in output of iostat command", "deviceName", stat.DeviceName)
```

Comment:
This message will pop up every second if there are more devices than the maximum number. It would be good to register the ignored devices and only log it will be ignored once.